pkg/planner/core/logical_plan_builder.go

```diff
@@ -3835,13 +3835,33 @@ func (b *PlanBuilder) buildSelect(ctx context.Context, sel *ast.SelectStmt) (p b
 	}
 	l := sel.LockInfo
 	if l != nil && l.LockType != ast.SelectLockNone {
-		for _, tName := range l.Tables {
+		var tableList []*ast.TableName
+		var isExplicitSetTablesNames bool
+		if len(l.Tables) == 0 {
+			nodeW := resolve.NewNodeWWithCtx(sel.From, b.resolveCtx)
+			tableList = ExtractTableList(nodeW, false)
+		} else {
+			isExplicitSetTablesNames = true
+			tableList = l.Tables
+		}
+		for _, tName := range tableList {
 			// CTE has no *model.HintedTable, we need to skip it.
 			tNameW := b.resolveCtx.GetTableName(tName)
 			if tNameW == nil {
 				continue
 			}
-			b.ctx.GetSessionVars().StmtCtx.LockTableIDs[tNameW.TableInfo.ID] = struct{}{}
+			if isExplicitSetTablesNames {
+				b.ctx.GetSessionVars().StmtCtx.LockTableIDs[tNameW.TableInfo.ID] = struct{}{}
```

Comment:
Can you add comment to explain why we don't need to update `LockTableIDs` for implicit SELECT FOR UPDATE? I didn't understand this logic 😂

Reply:
It's for the stmt like `select * from t, t1 where t.a = t1.a for update of t` which only locks table `t` instead of all tables. I will add it into comments.

Reply:
I didn't find where the locking of "(instead of) all tables" happens. 😂  It's deleted as the old line 3844

Reply:
If `LockTableIDs` map is empty, it will lock all records from all tables. Besides, it will only lock the metioned in `of` part.